common/SmokeTests/SmokeTest/Program.cs

```diff
@@ -1,37 +1,214 @@
+    // constructor is found during the scan, take note for later use.
+
+    var hasParameterlessConstructor = false;
+
+    try
+    {
+        foreach (var member in type.GetMembers(BindingFlags.Public | BindingFlags.Instance))
+        {
+            if ((member is ConstructorInfo constructor) && (constructor.GetParameters().Length == 0))
+            {
+                hasParameterlessConstructor = true;
+            }
+        }
+
+        // If the type is non-generic, non-abstract, and has a parameterless constructor, it should be constructed.
+
+        if ((!type.ContainsGenericParameters) && (!type.IsAbstract) && (hasParameterlessConstructor))
+        {
+            try
+            {
+                _ = Activator.CreateInstance(type);
+            }
+            catch (TargetInvocationException ex) when (ex.InnerException is NotImplementedException)
+            {
+                // Expected; this occurs when the parameterless constructor is present but has been marked obsolete.  Since we do
+                // not report types without a parameterless constructor, ignore the type and do not log.
+            }
+            catch (TargetInvocationException ex) when (ex.InnerException is FileNotFoundException fileEx)
+            {
+                // Expected; this indicates that a library is not available on a platform.  For
+                // example, System.Windows.Forms on macOS.
+
+                LogInformation($"\t\t[{ fileEx.FileName }] needed for type [{ type.FullName }] was not found.  This is expected for some platform-specific types, such as those in System.Windows.Forms when running on a non-Windows platform.");
+            }
+            catch (TargetInvocationException ex) when (ex.InnerException is PlatformNotSupportedException)
+            {
+                // Expected; this indicates that a type is not available on a platform.  For example,
+                // Windows Principal-related types on Linux.
+
+                LogInformation($"\t\t[{ type.FullName }] is not available on this platform.  This is expected for some platform-specific types.");
+            }
+        }
+    }
+    catch (Exception ex)
+    {
+        ReportError();
+        LogError($"{Environment.NewLine }[{ type.FullName }] could not be processed.  Error: { ex }{ Environment.NewLine }");
+    }
+}
+
+IEnumerable<Assembly> LoadAssemblies(Assembly rootAssembly, string assemblyFileMask)
+{
+    var processedAssemblies = new HashSet<string>();
+    var assembliesToProcess = new Stack<Assembly>();
+
+    // Start with the root assembly.
+
+    assembliesToProcess.Push(rootAssembly);
+
+    // Include any libraries referenced that have not been explicitly loaded; these should appear in
+    // the directory that the entry assembly is in.
+
+    foreach (var file in Directory.GetFiles(Path.GetDirectoryName(rootAssembly.Location), assemblyFileMask, SearchOption.AllDirectories))
```

Comment:
What is going to cause these assemblies to be in the output directory? I don't think a normal package reference will cause the dll to be copied to the output does it?

Reply:
It does; anything package reference gets pulled down during restore and included as part of the build output.   Unless a type in the assembly is used in the code, the assembly won't be loaded normally - so, unfortunately, the best means of discovery that we have for the references is to sniff the directory.